# Capítulo 2: Time Series

### Problema 1

a) La función de verosimilitud vendría dada por:
$$\mathcal{L} = \pi_{0,1} \times P_{12} \times P_{21} \times P_{12} \times P_{21}$$
Reemplazando se obtiene:
$$\mathcal{L} = 0.5 \times 0.1 \times 0.3 \times 0.1 \times 0.3$$

In [59]:
L = 0.5 * 0.1 * 0.3 * 0.1 * 0.3;
L

0.00045

b) La función de verosimilitud vendría dada por:
$$\mathcal{L} = \pi_{0,1} \times P_{11} \times P_{11} \times P_{11} \times P_{11}$$
Reemplazando se obtiene:
$$\mathcal{L} = 0.5 \times 0.9^4$$

In [60]:
L = 0.5 * 0.9^4;
L

0.32805000000000006

c) La función de verosimilitud vendría dada por:
$$\mathcal{L} = \pi_{0,2} \times P_{22} \times P_{22} \times P_{22} \times P_{22}$$
Reemplazando se obtiene:
$$\mathcal{L} = 0.5 \times 0.7^4$$

In [61]:
L = 0.5 * 0.7^4;
L

0.12004999999999996

### Problema 2

Podemos tratar de averiguar cuál es la expresión que representa $E[y^2_{t + 1} | x_t]$. En primer lugar tenemos:
$$E[y^2_{t + 1} | x_t = e_1] = (\bar{y}'e_1)^2 P_{11} + (\bar{y}'e_2)^2 P_{12}$$
Luego tenemos:
$$E[y^2_{t + 1} | x_t = e_2] = (\bar{y}'e_1)^2 P_{21} + (\bar{y}'e_2)^2 P_{22}$$

Entonces tendríamos que:
$$E[y^2_{t + 1} | x_t] = P \left[\begin{array}{c} 1 \\
                                                25\end{array}\right]$$

De manera análoga tendríamos que:
$$E[y_{t + 1} | x_t] = P \left[\begin{array}{c} 1 \\ 5 \end{array}\right]$$

Entonces se tendría el sistema:
$$\left[\begin{array}{cc} 1.8 & 5.8 \\ 3.4 & 15.4 \end{array}\right] = P \left[\begin{array}{cc} 1 & 1 \\ 5 & 25\end{array}\right]$$
$$J = Ph$$

La solución es única si el rango de la matriz $h'$ es completo o sea 2.

In [62]:
using LinearAlgebra

In [63]:
h = [1 1; 5 25]; J = [1.8 5.8; 3.4 15.4];

In [64]:
rank(h')

2

Entonces la matrix $P$ es única. La solución viene dada por:

In [65]:
X = h' \ J';
P = X'

2×2 Adjoint{Float64,Array{Float64,2}}:
 0.8  0.2
 0.4  0.6

### Problema 3

a) Sabemos que:
$$v_i = E\left[\sum^\infty_{t = 0} \beta^t u(c_t) | x_0 = \bar{e}_i\right] = \sum^\infty_{t = 0} E[\beta^t u(c_t) | x_0 = \bar{e}_i]$$

Por lo tanto, nos interesa saber que expresión representa $E[\beta^t u(c_t) | x_0 = \bar{e}_i]$. Tenemos que:
$$E[\beta^t u(c_t) | x_0 = \bar{e}_i] = \beta^t u(\bar{c}'\bar{e}_1)P^{(t)}_{i1} + \beta^t u(\bar{c}'\bar{e}_2)P^{(t)}_{i2} + \dots +\beta^t u(\bar{c}'\bar{e}_n)P^{(t)}_{in}$$ 

Entonces:
$$E[\beta^t u(c_t) | x_0 = \bar{e}_i] = \beta^t u_1 P^{(t)}_{i1} + \beta^t u_2 P^{(t)}_{i2} + \dots + \beta^t u_n P^{(t)}_{in}$$
Por lo que:
$$E[\beta^t u(c_t) | x_0] = \beta^t P^t u$$ 

Con esto se tiene que:
$$v = \sum^\infty_{t = 0} E[\beta^t u(c_t) | x_0] = \sum^\infty_{t = 0} \beta^t P^t u = (I - \beta P)^{-1}u$$

En cuanto a $V$ tenemos que:
$$V = E[v] = \pi'_0 v = \sum^n_{i = 1} \pi_{0,i} v_i$$

b) Construyamos funciones para: utilidad, $v$ y $V$:

In [66]:
u(c, γ) = c^(1 - γ) / (1 - γ) 

u (generic function with 1 method)

In [67]:
v(P, β, u) = (I - β * P)^(-1) * u

v (generic function with 1 method)

In [68]:
V(π_0, v) = π_0' * v

V (generic function with 1 method)

Construyamos el vector de utilidades en función del vector de consumo $\bar{c} = \left[\begin{array}{c} 1 \\
                                                                                                        5\end{array}\right]$ y $\gamma = 2.5$

In [69]:
c = [1, 5]; γ = 2.5;
util = u.(c, γ)

2-element Array{Float64,1}:
 -0.6666666666666666
 -0.05962847939999439

Consideremos el primer proceso:

In [70]:
P = [1 0; 0 1]; π_0 = [0.5, 0.5];

In [71]:
v_val = v(P, 0.95, util);
V_val = V(π_0, v_val)

-7.262951460666605

Consideremos el segundo proceso:

In [72]:
P = [0.5 0.5; 0.5 0.5]; π_0 = [0.5, 0.5];

In [73]:
v_val = v(P, 0.95, util);
V_val = V(π_0, v_val)

-7.262951460666601

Ambos procesos arrojan valores identicos $V_1 = V_2 = -7.26$ por lo que el consumidor es indiferente. Ahora analizemos el caso de $\gamma = 4$.

In [74]:
util = u.(c, 4.0)

2-element Array{Float64,1}:
 -0.3333333333333333
 -0.0026666666666666666

Consideremos el primer proceso:

In [75]:
P = [1 0; 0 1]; π_0 = [0.5, 0.5];
v_val = v(P, 0.95, util);
V_val = V(π_0, v_val)

-3.359999999999997

Ahora el segundo proceso:

In [76]:
P = [0.5 0.5; 0.5 0.5]; π_0 = [0.5, 0.5];
v_val = v(P, 0.95, util);
V_val = V(π_0, v_val)

-3.359999999999996

Nuevamente ambos procesos arrojan valores idénticos $V_1 = V_2 = -3.36$ por lo que el consumidor es indiferente.

c) Para el primer proceso la función de verosimilitud sería:
$$Prob(data | M_1) = \pi_{0,1} \times P^9_{11} = 0.5 \times 1^9$$

In [77]:
L_1 = 0.5 * 1^9

0.5

Para el segundo proceso la función de verosimilitud sería:
$$Prob(data | M_2) = \pi_{0,1} \times P^9_{11} = 0.5 \times 0.5^9 = 0.5^{10}$$

In [78]:
L_2 = 0.5^(10)

0.0009765625

d) Como las probabilidades iniciales de ambos modelos son idénticas tenemos que: $Prob(M_1) = Prob(M_2)$; por lo que:
$$Prob(M_1 | data) = \frac{Prob(data | M_1)}{Prob(data | M_1) + Prob(data | M_2)} = \frac{0.5}{0.5 + 0.001}$$

In [79]:
p1_n = L_1 / (L_1 + L_2)

0.9980506822612085

Para el caso del segundo proceso tenemos:
$$Prob(M_2 | data) = \frac{Prob(data | M_2)}{Prob(data | M_1) + Prob(data | M_2)} = \frac{0.001}{0.5 + 0.001}$$

In [80]:
p2_n = L_2 / (L_1 + L_2)

0.001949317738791423

e) Para la segunda muestra de datos, tenemos que:
$$Prob(data | M_1) = \pi_{0,1} \times P^4_{12} \times P^3_{21} \times P^2_{22} = 0.5 \times 0^4 \times 0^3 \times 1^2$$ 

In [81]:
L_1 = 0.5 * 0^4 * 0^3 * 1^2

0.0

Para el segundo proceso tenemos:
$$Prob(data | M_2) = \pi_{0,1} \times P^4_{12} \times P^3_{21} \times P^2_{22} = 0.5^{10}$$

In [82]:
L_2 = 0.5^(10)

0.0009765625

Tendríamos entonces que:
$$Prob(M_1 | data) = \frac{0}{0 + 0.001}$$

In [83]:
p1_n = L_1 / (L_1 + L_2)

0.0

Para el caso del segundo modelo:
$$Prob(M_2 | data) = \frac{0.001}{0 + 0.001}$$

In [84]:
p2_n = L_2 / (L_1 + L_2)

1.0

### Problema 4

a) Tenemos el siguiente modelo univariante:
$$y_{t + 1} = \alpha + \sum^4_{j = 1} \rho_j y_{t + 1 - j} + c w_{t + 1}$$

Podemos reesxpresarlo como:
$$\left[\begin{array}{c} y_{t + 1} \\ y_t \\ y_{t - 1} \\ y_{t - 2} \\ 1\end{array}\right] = \left[\begin{array}{ccccc} \rho_1 & \rho_2 & \rho_3 & \rho_4 & \alpha \\ 1 & 0 & 0 & 0 & 0 \\ 0 & 1 & 0 & 0 & 0 \\ 0 & 0 & 1 & 0 & 0 \\ 0 & 0 & 0 & 0 & 1 \end{array}\right] \left[ \begin{array}{c} y_t \\ y_{t - 1} \\ y_{t - 2} \\ y_{t - 3} \\ 1 \end{array}\right] + \left[\begin{array}{c} c \\ 0 \\ 0 \\ 0 \\ 0 \end{array}\right] w_{t + 1}$$
$$\mathbf{Y}_{t + 1} = \mathbf{A}\mathbf{Y}_t + \mathbf{C}w_{t + 1}$$

b) Para verificar que el proceso es estacionario, se debe verificar que los autovalores de la matriz: 
$$\mathbf{A}_0 = \left[ \begin{array}{cccc} \rho_1 & \rho_2 & \rho_3 & \rho_4 \\ 1 & 0 & 0 & 0 \\ 0 & 1 & 0 & 0 \\ 0 & 0 & 1 & 0 \end{array}\right]$$
Sean en módulo menor a 1.

i) 

In [85]:
ρ = [1.2 -0.3 0 0]; μ = 10; c = 1;

In [86]:
A_0 = [ρ; 1 0 0 0; 0 1 0 0; 0 0 1 0];
evals, evec = eigen(A_0);
abs.(evals)

4-element Array{Float64,1}:
 0.0
 0.0
 0.35505102572168223
 0.8449489742783177

Como todos los autovalores son inferiores a la unidad en módulo, tenemos que el proceso es estacionario. La media del proceso vendría dada por el autovector asociado al autovalor unitario de la matriz $\mathbf{A}$:

In [87]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
evals, evec = eigen(A);
abs.(evals)

5-element Array{Float64,1}:
 0.0
 0.0
 0.35505102572168223
 0.8449489742783177
 1.0

In [88]:
evec

5×5 Array{Float64,2}:
 0.0   0.0           -0.0418473  -0.375014  0.499376
 0.0   0.0           -0.117863   -0.44383   0.499376
 0.0   5.01042e-292  -0.33196    -0.525274  0.499376
 1.0  -1.0           -0.934965   -0.621664  0.499376
 0.0   0.0            0.0         0.0       0.0499376

In [89]:
μ_Y = evec[:, 5] ./ evec[5,5]

5-element Array{Float64,1}:
 10.000000000000005
 10.000000000000007
 10.000000000000007
 10.000000000000007
  1.0

Como $y_t = J\mathbf{Y}_t$ donde $J = [\begin{array}{ccccc} 1 & 0 & 0 & 0 & 0 \end{array}]$. Tenemos que $\mu_y = J\mu_Y$. Entonces:

In [90]:
J = [1 0 0 0 0]; 
μ_y = J * μ_Y

1-element Array{Float64,1}:
 10.000000000000005

Entonces la media del proceso $\mu_y = 10$. Ahora para la varianza del proceso $C_Y(0)$ utilizamos la ecuación de lyapunov:
$$C_Y(0) = \mathbf{A} C_Y(0) \mathbf{A}' + \mathbf{CC'}$$

In [91]:
C = [c, 0, 0, 0, 0];
C_C = C * C';

Codificaremos la función utilizada por Thomas Sargent en MATLAB:

In [92]:
function doublej(a1, b1)
    A_0 = a1
    G_0 = b1
    diff = 5; ijk = 1;
    while diff > 1e-15 
       A_1 = A_0 * A_0
       G_1 = G_0 + A_0 * G_0 * A_0'
       aux = maximum(maximum(abs.(G_1 - G_0), dims = 1), dims = 2)
       diff = aux[1] 
       G_0 = G_1
       A_0 = A_1
       ijk += 1 
    end
    return G_0
end

doublej (generic function with 1 method)

In [93]:
X = doublej(A, C_C)

5×5 Array{Float64,2}:
 7.42857  6.85714  6.0      5.14286  0.0
 6.85714  7.42857  6.85714  6.0      0.0
 6.0      6.85714  7.42857  6.85714  0.0
 5.14286  6.0      6.85714  7.42857  0.0
 0.0      0.0      0.0      0.0      0.0

Tenemos que $C_y(0) = J C_Y(0) J'$; por lo tanto: 

In [94]:
C_y = J * X * J'

1×1 Array{Float64,2}:
 7.428571428571427

Entonces $Var_y = 7.43$.

ii)

In [95]:
ρ = [1.2 -0.3 0 0]; μ = 10; c = 2;

In [96]:
A_0 = [ρ; 1 0 0 0; 0 1 0 0; 0 0 1 0];
evals, evec = eigen(A_0);
abs.(evals)

4-element Array{Float64,1}:
 0.0
 0.0
 0.35505102572168223
 0.8449489742783177

El proceso es estacionario. Calculamos la media del proceso como el autovector asociado al autovalor unitario de la matriz $\mathbf{A}$.

In [97]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
evals, evec = eigen(A);
abs.(evals)

5-element Array{Float64,1}:
 0.0
 0.0
 0.35505102572168223
 0.8449489742783177
 1.0

In [98]:
μ_Y = evec[:, 5] ./ evec[5,5]

5-element Array{Float64,1}:
 10.000000000000005
 10.000000000000007
 10.000000000000007
 10.000000000000007
  1.0

In [99]:
J = [1 0 0 0 0]; 
μ_y = J * μ_Y

1-element Array{Float64,1}:
 10.000000000000005

La media del proceso $\mu_y = 10$. Ahora calculamos la varianza:

In [100]:
C = [c, 0, 0, 0, 0];
C_C = C * C';
X = doublej(A, C_C)

5×5 Array{Float64,2}:
 29.7143  27.4286  24.0     20.5714  0.0
 27.4286  29.7143  27.4286  24.0     0.0
 24.0     27.4286  29.7143  27.4286  0.0
 20.5714  24.0     27.4286  29.7143  0.0
  0.0      0.0      0.0      0.0     0.0

In [101]:
C_y = J * X * J'

1×1 Array{Float64,2}:
 29.714285714285708

Entonces $Var_y = 29.71$.

iii)

In [102]:
ρ = [.9 0 0 0]; μ = 5; c = 1;

In [103]:
A_0 = [ρ; 1 0 0 0; 0 1 0 0; 0 0 1 0];
evals, evec = eigen(A_0);
abs.(evals)

4-element Array{Float64,1}:
 0.0
 0.0
 0.0
 0.9

El proceso es estacionario. Calculamos la media del proceso:

In [104]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
evals, evec = eigen(A);
abs.(evals)

5-element Array{Float64,1}:
 0.0
 0.0
 0.0
 0.9
 1.0

In [105]:
μ_Y = evec[:, 5] ./ evec[5,5]

5-element Array{Float64,1}:
 5.0
 5.0
 5.0
 5.0
 1.0

In [106]:
J = [1 0 0 0 0]; 
μ_y = J * μ_Y

1-element Array{Float64,1}:
 5.0

La media del proceso es $\mu_y = 5$. Ahora calculamos la varianza:

In [107]:
C = [c, 0, 0, 0, 0];
C_C = C * C';
X = doublej(A, C_C)

5×5 Array{Float64,2}:
 5.26316  4.73684  4.26316  3.83684  0.0
 4.73684  5.26316  4.73684  4.26316  0.0
 4.26316  4.73684  5.26316  4.73684  0.0
 3.83684  4.26316  4.73684  5.26316  0.0
 0.0      0.0      0.0      0.0      0.0

In [108]:
C_y = J * X * J'

1×1 Array{Float64,2}:
 5.263157894736844

Entonces $Var_y = 5.26$.

iv) 

In [109]:
ρ = [.2 0 0 .5]; μ = 5; c = 1;

In [110]:
A_0 = [ρ; 1 0 0 0; 0 1 0 0; 0 0 1 0];
evals, evec = eigen(A_0);
abs.(evals)

4-element Array{Float64,1}:
 0.7950219427055532
 0.8379288196679984
 0.8379288196679984
 0.8957289964623478

El proceso es estacionario. Ahora calculamos la media del proceso:

In [111]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
evals, evec = eigen(A);
abs.(evals)

5-element Array{Float64,1}:
 0.7950219427055532
 0.8379288196679984
 0.8379288196679984
 0.8957289964623478
 1.0

In [112]:
μ_Y = evec[:, 5] ./ evec[5,5]

5-element Array{Complex{Float64},1}:
  5.000000000000029 + 0.0im
 5.0000000000000275 + 0.0im
   5.00000000000003 + 0.0im
   5.00000000000004 + 0.0im
 0.9999999999999999 + 0.0im

In [113]:
J = [1 0 0 0 0]; 
μ_y = J * μ_Y

1-element Array{Complex{Float64},1}:
 5.000000000000029 + 0.0im

La media del proceso es $\mu_y = 5$. Ahora calculamos la varianza:

In [114]:
C = [c, 0, 0, 0, 0];
C_C = C * C';
X = doublej(A, C_C)

5×5 Array{Float64,2}:
 1.4764    0.415887  0.166355  0.241214  0.0
 0.415887  1.4764    0.415887  0.166355  0.0
 0.166355  0.415887  1.4764    0.415887  0.0
 0.241214  0.166355  0.415887  1.4764    0.0
 0.0       0.0       0.0       0.0       0.0

In [115]:
C_y = J * X * J'

1×1 Array{Float64,2}:
 1.4763984196298612

La varianza del proceso es $Var_y = 1.47$.

v)

In [116]:
ρ = [.8 .3 0 0]; μ = 5; c = 1;

In [117]:
A_0 = [ρ; 1 0 0 0; 0 1 0 0; 0 0 1 0];
evals, evec = eigen(A_0);
abs.(evals)

4-element Array{Float64,1}:
 0.2782329983125267
 0.0
 0.0
 1.078232998312527

El proceso no es estacionario por lo que no podemos calcular la media del proceso ni la varianza del proceso.

c) Lo que se busca es:
$$y_{t + 5} = \gamma_0 + \sum^3_{j = 0}h_j y_{t - j} + w_{t + 5}$$
Para los incisos i), ii), iii) y iv) los procesos son estacionarios y nos enfocamos sólo en ellos; entonces tenemos:
$$\mu = \gamma_0 + \sum^3_{j = 0}h_j \mu$$
Por lo que:
$$y_{t + 5} - \mu = \sum^3_{j = 0}h_j (y_{t - j} - \mu) + w_{t + 5}$$
Lo que vendría a ser una proyección lineal de $y_{t + 5} - \mu$ en $\mathbf{X}_t = \left[\begin{array}{c} y_t - \mu \\ y_{t - 1} - \mu \\ y_{t - 2} - \mu \\ y_{t - 3} - \mu \end{array}\right]$. Por lo tanto:
$$h = E((y_{t + 5} - \mu)\mathbf{X}_t')[E(\mathbf{X_t X_t'})]^{-1}$$

La expresión $E((y_{t + 5} - \mu)\mathbf{X}_t')$ sería igual a:
$$E((y_{t + 5} - \mu)\mathbf{X}_t') = JE(\mathbf{X}_{t + 5}\mathbf{X}_t') = J C_Y(5) = J \mathbf{A}^5 C_Y(0)$$

De esta última expresión sólo tomamos los 4 primeros valores. Para calcular $E(\mathbf{X_tX_t'})$ tomamos las 4 primeras filas y 4 primeras columnas de la matriz $C_Y(0)$. Finalmente para calcular $\gamma_0 = \mu(1 - \sum^3_{j = 0}h_j)$.

i)

In [118]:
ρ = [1.2 -.3 0 0]; μ = 10; c = 1;

In [119]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
C = [c, 0, 0, 0, 0];
C_C = C * C';
C_Y = doublej(A, C_C)

5×5 Array{Float64,2}:
 7.42857  6.85714  6.0      5.14286  0.0
 6.85714  7.42857  6.85714  6.0      0.0
 6.0      6.85714  7.42857  6.85714  0.0
 5.14286  6.0      6.85714  7.42857  0.0
 0.0      0.0      0.0      0.0      0.0

In [120]:
J = [1 0 0 0 0]; 
E_y_X = J * A^5 * C_Y;
E_y_X = E_y_X[1, 1:4];
E_X_X = C_Y[1:4, 1:4];
h = E_y_X' * (E_X_X)^(-1)

1×4 Adjoint{Float64,Array{Float64,1}}:
 0.73872  -0.26028  -3.55271e-15  1.33227e-15

In [121]:
γ_0 = μ * (1 - sum(h))

5.215600000000014

ii)

In [123]:
ρ = [1.2 -.3 0 0]; μ = 10; c = 2;

In [124]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
C = [c, 0, 0, 0, 0];
C_C = C * C';
C_Y = doublej(A, C_C)

5×5 Array{Float64,2}:
 29.7143  27.4286  24.0     20.5714  0.0
 27.4286  29.7143  27.4286  24.0     0.0
 24.0     27.4286  29.7143  27.4286  0.0
 20.5714  24.0     27.4286  29.7143  0.0
  0.0      0.0      0.0      0.0     0.0

In [125]:
J = [1 0 0 0 0]; 
E_y_X = J * A^5 * C_Y;
E_y_X = E_y_X[1, 1:4];
E_X_X = C_Y[1:4, 1:4];
h = E_y_X' * (E_X_X)^(-1)

1×4 Adjoint{Float64,Array{Float64,1}}:
 0.73872  -0.26028  -3.55271e-15  1.33227e-15

In [126]:
γ_0 = μ * (1 - sum(h))

5.215600000000014

iii)

In [128]:
ρ = [.9 0 0 0]; μ = 5; c = 1;

In [129]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
C = [c, 0, 0, 0, 0];
C_C = C * C';
C_Y = doublej(A, C_C)

5×5 Array{Float64,2}:
 5.26316  4.73684  4.26316  3.83684  0.0
 4.73684  5.26316  4.73684  4.26316  0.0
 4.26316  4.73684  5.26316  4.73684  0.0
 3.83684  4.26316  4.73684  5.26316  0.0
 0.0      0.0      0.0      0.0      0.0

In [130]:
J = [1 0 0 0 0]; 
E_y_X = J * A^5 * C_Y;
E_y_X = E_y_X[1, 1:4];
E_X_X = C_Y[1:4, 1:4];
h = E_y_X' * (E_X_X)^(-1)

1×4 Adjoint{Float64,Array{Float64,1}}:
 0.59049  8.88178e-16  -8.88178e-16  0.0

In [131]:
γ_0 = μ * (1 - sum(h))

2.0475500000000024

iv)

In [132]:
ρ = [.2 0 0 .5]; μ = 5; c = 1;

In [133]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
C = [c, 0, 0, 0, 0];
C_C = C * C';
C_Y = doublej(A, C_C)

5×5 Array{Float64,2}:
 1.4764    0.415887  0.166355  0.241214  0.0
 0.415887  1.4764    0.415887  0.166355  0.0
 0.166355  0.415887  1.4764    0.415887  0.0
 0.241214  0.166355  0.415887  1.4764    0.0
 0.0       0.0       0.0       0.0       0.0

In [134]:
J = [1 0 0 0 0]; 
E_y_X = J * A^5 * C_Y;
E_y_X = E_y_X[1, 1:4];
E_X_X = C_Y[1:4, 1:4];
h = E_y_X' * (E_X_X)^(-1)

1×4 Adjoint{Float64,Array{Float64,1}}:
 0.20032  0.02  0.004  0.2508

In [135]:
γ_0 = μ * (1 - sum(h))

2.6244

d) Consideramos unicamente los casos i, ii, iii, y iv donde los procesos son estacionarios. Tenemos que:
$$E_t \sum^\infty_{k = 0} .95^k y_{t + k} = E_t \sum^\infty_{k = 0} .95^k J \mathbf{Y}_{t + k} = J \sum^\infty_{k = 0} .95^k A^k \mathbf{Y}_t = J (\mathbf{I} - .95 \mathbf{A})^{-1}\mathbf{Y}_t$$
Por lo tanto:
$$[\begin{array}{ccccc} h_0 & h_1 & h_2 & h_3 & \gamma_0 \end{array}] = J(\mathbf{I} - .95 \mathbf{A})^{-1}$$

i)

In [136]:
ρ = [1.2 -.3 0 0]; μ = 10; c = 1;

In [137]:
J = [1 0 0 0 0];
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
h = J * (I - 0.95 * A)^(-1)

1×5 Array{Float64,2}:
 7.64818  -2.17973  0.0  0.0  145.315

ii)

In [138]:
ρ = [1.2 -.3 0 0]; μ = 10; c = 2;

In [139]:
J = [1 0 0 0 0];
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
h = J * (I - 0.95 * A)^(-1)

1×5 Array{Float64,2}:
 7.64818  -2.17973  0.0  0.0  145.315

iii)

In [140]:
ρ = [.9 0 0 0]; μ = 5; c = 1;

In [141]:
J = [1 0 0 0 0];
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
h = J * (I - 0.95 * A)^(-1)

1×5 Array{Float64,2}:
 6.89655  -2.22045e-16  -2.22045e-16  0.0  65.5172

iv)

In [142]:
ρ = [.2 0 0 .5]; μ = 5; c = 1;

In [143]:
J = [1 0 0 0 0];
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
h = J * (I - 0.95 * A)^(-1)

1×5 Array{Float64,2}:
 2.48295  1.06441  1.12043  1.1794  70.7641

La k-ésima autocovarianza viene dada por:
$$E(y_t - \mu)(y_{t - k} - \mu) = C_y(k) = JC_Y(k)J' = J A^k C_Y(0) J'$$

Consideramos sólo los casos estacionarios i, ii, iii y iv.

i)

In [144]:
ρ = [1.2 -.3 0 0]; μ = 10; c = 1;

In [145]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
C = [c, 0, 0, 0, 0];
C_C = C * C';
C_Y = doublej(A, C_C)

5×5 Array{Float64,2}:
 7.42857  6.85714  6.0      5.14286  0.0
 6.85714  7.42857  6.85714  6.0      0.0
 6.0      6.85714  7.42857  6.85714  0.0
 5.14286  6.0      6.85714  7.42857  0.0
 0.0      0.0      0.0      0.0      0.0

In [146]:
J = [1 0 0 0 0];
C_y_1 = J * A * C_Y * J';
C_y_5 = J * A^5 * C_Y * J';
C_y_10 = J * A^10 * C_Y * J';
autocov = [C_y_1 C_y_5 C_y_10]

1×3 Array{Float64,2}:
 6.85714  3.70286  1.59758

ii)

In [147]:
ρ = [1.2 -.3 0 0]; μ = 10; c = 2;

In [148]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
C = [c, 0, 0, 0, 0];
C_C = C * C';
C_Y = doublej(A, C_C)

5×5 Array{Float64,2}:
 29.7143  27.4286  24.0     20.5714  0.0
 27.4286  29.7143  27.4286  24.0     0.0
 24.0     27.4286  29.7143  27.4286  0.0
 20.5714  24.0     27.4286  29.7143  0.0
  0.0      0.0      0.0      0.0     0.0

In [149]:
J = [1 0 0 0 0];
C_y_1 = J * A * C_Y * J';
C_y_5 = J * A^5 * C_Y * J';
C_y_10 = J * A^10 * C_Y * J';
autocov = [C_y_1 C_y_5 C_y_10]

1×3 Array{Float64,2}:
 27.4286  14.8114  6.39032

iii)

In [150]:
ρ = [.9 0 0 0]; μ = 5; c = 1;

In [151]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
C = [c, 0, 0, 0, 0];
C_C = C * C';
C_Y = doublej(A, C_C)

5×5 Array{Float64,2}:
 5.26316  4.73684  4.26316  3.83684  0.0
 4.73684  5.26316  4.73684  4.26316  0.0
 4.26316  4.73684  5.26316  4.73684  0.0
 3.83684  4.26316  4.73684  5.26316  0.0
 0.0      0.0      0.0      0.0      0.0

In [152]:
J = [1 0 0 0 0];
C_y_1 = J * A * C_Y * J';
C_y_5 = J * A^5 * C_Y * J';
C_y_10 = J * A^10 * C_Y * J';
autocov = [C_y_1 C_y_5 C_y_10]

1×3 Array{Float64,2}:
 4.73684  3.10784  1.83515

iv)

In [155]:
ρ = [.2 0 0 .5]; μ = 5; c = 1;

In [156]:
α = μ * (1 - sum(ρ));
A = [ρ α; 1 0 0 0 0; 0 1 0 0 0; 0 0 1 0 0; 0 0 0 0 1];
C = [c, 0, 0, 0, 0];
C_C = C * C';
C_Y = doublej(A, C_C)

5×5 Array{Float64,2}:
 1.4764    0.415887  0.166355  0.241214  0.0
 0.415887  1.4764    0.415887  0.166355  0.0
 0.166355  0.415887  1.4764    0.415887  0.0
 0.241214  0.166355  0.415887  1.4764    0.0
 0.0       0.0       0.0       0.0       0.0

In [157]:
J = [1 0 0 0 0];
C_y_1 = J * A * C_Y * J';
C_y_5 = J * A^5 * C_Y * J';
C_y_10 = J * A^10 * C_Y * J';
autocov = [C_y_1 C_y_5 C_y_10]

1×3 Array{Float64,2}:
 0.415887  0.365232  0.131579